src/main.rs

```diff
@@ -0,0 +1,253 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#[macro_use(value_t)]
+extern crate clap;
+
+use clap::{App, AppSettings, Arg, SubCommand};
+use manticore::io::StdWrite;
+use manticore::protocol;
+use manticore::protocol::firmware_version;
+use manticore::protocol::wire::FromWire;
+use manticore::protocol::wire::ToWire;
+use manticore::protocol::CommandType;
+use manticore::protocol::Header;
+use serde_json;
+use std::fs::OpenOptions;
+use std::io::prelude::*;
+use std::io::BufReader;
+
+/// Deserializes a message in JSON format from `reader` and then serializes the message in wire
+/// format to `writer`.
+fn from_json_to_wire<'a, T, R, W>(reader: R, writer: W)
+where
+    T: ToWire + serde::de::Deserialize<'a>,
+    R: std::io::Read,
+    W: manticore::io::Write,
+{
+    let mut de = serde_json::Deserializer::from_reader(reader);
+    let msg = T::deserialize(&mut de).expect("failed to deserialize JSON");
+    msg.to_wire(writer).expect("failed to write request");
+}
+
+/// Deserializes a message in JSON format from `input_file` and then serializes a header + message
+/// in wire format to.`output_file`. Uses `cmd_type` and `is_request` to determine the message type
+///  since the JSON format does not include the message header.
+fn from_json(
+    cmd_type: CommandType,
+    is_request: bool,
+    input_file: &str,
+    output_file: &str,
+) {
+    let mut output = OpenOptions::new()
+        .write(true)
+        .truncate(true)
+        .create(true)
+        .open(&output_file)
+        .expect("failed to open output file");
+
+    let input = OpenOptions::new()
+        .read(true)
+        .open(&input_file)
+        .expect("failed to open input file");
+    let input_reader = BufReader::new(input);
+
+    let mut stdwrite = StdWrite(&mut output);
+
+    let header = Header {
+        is_request: is_request,
+        command: cmd_type,
+    };
+    header
+        .to_wire(&mut stdwrite)
+        .expect("failed to write header");
+
+    if is_request {
```

Comment:
You could write this as:
```
match (is_request, command_type) {
  (true, CommandType::FirmwareVersion) => ...,
  (false, CommandType::FirmwareVersion) => ...,
  _ => ...,
}
```

Reply:
Ah yes, good point! I hadn't thought of this. Nice trick to know!